added contacts recycler view

cValentyna opened the pull request:
added contacts recycler view;
implemented data initialization via viewmodel;
added image load via url;

cValentyna committed:
level 2 part 1
added contacts recycler view;
implemented data initialization via viewmodel;
added image load via url;

cValentyna committed:
updated libraries - partly

kir1hime commented on app/src/main/java/com/project/myapp/screens/contacts/ContactActivity.kt:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    private fun collectUserList() {
+        lifecycleScope.launch {
```
You didn't consider a lifecycle of Avtitvity, read about repeatOnLifecycle

cValentyna replied:
I read about it after your recommendation. It was [changed](https://github.com/cValentyna/myAndroidProject/pull/18/commits/83214a63246096d131dfa5d711549869f905fa75#diff-df8e14d58a9d99e55bff0b87a86bca4aa2126938b65dfc678e78eaceb10d560cR56-R60).

kir1hime commented on app/src/main/java/com/project/myapp/screens/contacts/ItemDecorator.kt:
```diff
@@ -0,0 +1,55 @@
+import android.view.View
+import androidx.recyclerview.widget.RecyclerView
+
+class ItemDecorator(
```
It’s Ok if you wanted to practice implementing item views this way. But if that wasn’t, and you just didn’t think about it, you can do it in xml of items and contacts. Here you only need to set offsets

cValentyna replied:
it was changed [here](https://github.com/cValentyna/myAndroidProject/pull/18/commits/06404ad6e2ffcae480d6591de146bc2351b503ec).

kir1hime commented on app/src/main/res/layout/contact_item.xml:
```diff
@@ -0,0 +1,65 @@
+<!--        android:src="@drawable/person_icon_png_transparent_15"-->
+<!--        app:shapeAppearanceOverlay="@style/shapeAppearance_image" />-->
+
+    <ImageView
```
It is better to use AppCompat components

kir1hime commented on app/src/main/res/layout/activity_contacts.xml:
```diff
@@ -0,0 +1,87 @@
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintBottom_toBottomOf="@id/contact_view_top_part" />
+
+    <androidx.recyclerview.widget.RecyclerView
```
What are biases here for?

cValentyna replied:
thank you for pointing this out. It was just duplicating of other settings. my fault.

kir1hime commented on app/src/main/res/layout/activity_contacts.xml:
```diff
@@ -0,0 +1,87 @@
+
+    <androidx.recyclerview.widget.RecyclerView
+        android:id="@+id/recyclerViewContacts"
+        android:layout_width="0dp"
```
Maby set match_match parent and delete    
app:layout_constraintLeft_toLeftOf="parent"
app:layout_constraintRight_toRightOf="parent"

cValentyna replied:
I completely agree with you.

kir1hime commented on app/src/main/res/layout/activity_contacts.xml:
```diff
@@ -0,0 +1,87 @@
+<?xml version="1.0" encoding="utf-8"?>
+<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
Use text formatting)

kir1hime commented on app/src/main/res/layout/activity_contacts.xml:
```diff
@@ -0,0 +1,87 @@
+        app:layout_constraintHorizontal_chainStyle="spread_inside"
+         />
+
+    <androidx.appcompat.widget.AppCompatTextView
```
For what center_vertical if you use wrap_content?

kir1hime commented on app/src/main/res/drawable/contact_find.xml:
```diff
@@ -0,0 +1,16 @@
+<vector xmlns:android="http://schemas.android.com/apk/res/android"
```
set more clear name)

cValentyna replied:
[Name](https://github.com/cValentyna/myAndroidProject/pull/18/commits/83214a63246096d131dfa5d711549869f905fa75#diff-df8e14d58a9d99e55bff0b87a86bca4aa2126938b65dfc678e78eaceb10d560cR56-R60:~:text=android%3Asrc%3D%22%40drawable/contact_find%22) was changed

kir1hime commented on app/src/main/java/com/project/myapp/screens/contacts/UsersRepository.kt:
```diff
@@ -0,0 +1,75 @@
+package com.project.myapp.screens.contacts
```
replace all classes that work with data into the data (or model) package. In sсreens package you need only those classes wich work with view

cValentyna replied:
[data classes](https://github.com/cValentyna/myAndroidProject/pull/18/commits/1e98b2705be92db6832bd4bf7c170fdc26db613f) were moved. For now, I decided not to move the viewmodel from the activity yet.

kir1hime replied:
ViewModel must be near Activity. It is right

kir1hime commented on app/src/main/java/com/project/myapp/screens/contacts/User.kt:
```diff
@@ -0,0 +1,8 @@
+package com.project.myapp.screens.contacts
```
to data (model) package

kir1hime commented on app/src/main/java/com/project/myapp/screens/contacts/ContactActivity.kt:
```diff
@@ -0,0 +1,55 @@
+        binding.root.initializeWindowInsetsHandling()
+    }
+
+    private fun initRecyclerView() {
```
Can see what is with() function, and use with(binding). It is very useful for me personally

cValentyna replied:
It was changed [here](https://github.com/cValentyna/myAndroidProject/pull/18/commits/83214a63246096d131dfa5d711549869f905fa75#diff-df8e14d58a9d99e55bff0b87a86bca4aa2126938b65dfc678e78eaceb10d560cR39-R51)

kir1hime replied:
May be simplify 

private fun initRecyclerView() = with(binding) {
recyclerViewContacts.addItemDecoration(ItemDecorator(resources.getDimensionPixelSize(R.dimen.gap_item)))
recyclerViewContacts.layoutManager = LinearLayoutManager(this@ContactActivity)
recyclerViewContacts.adapter = contactsAdapter
}

kir1hime commented on app/src/main/java/com/project/myapp/screens/contacts/UsersRepository.kt:
```diff
@@ -0,0 +1,75 @@
+        if (minLength == 0) return emptyList()
+
+        return List(minLength) { index ->
+            User(id = getNextId(), name = nameList[index], profession = professionList[index], photoUrl = photoList[index])
```
Maybe just id=index+1))

kir1hime commented:
Hilt modules also must be replaced into separate package (usually it is called di )

cValentyna committed:
level 2 part 2
correction according to the comments in the review

cValentyna committed:
level 2 part 2
correction according to the comments in the review.
changed packages, moved data classes  and AppModule to separate folders.

cValentyna committed:
level 2 part 2
correction according to the comments in the review.
moved part of item decorator logic to contact_item.xml

kir1hime commented on app/src/main/java/com/project/myapp/data/contacts/UsersRepository.kt:
```diff
@@ -0,0 +1,75 @@
+        }
+    }
+
+    private fun getNextId(): Int = ++nextId
```
delete unnecessary method and field

kir1hime commented on app/src/main/java/com/project/myapp/screens/contacts/ContactActivity.kt:
```diff
@@ -0,0 +1,56 @@
+    }
+
+    private fun initRecyclerView() {
+        with(binding) {
```
May be simplify

private fun initRecyclerView() = with(binding) {
recyclerViewContacts.addItemDecoration(ItemDecorator(resources.getDimensionPixelSize(R.dimen.gap_item)))
recyclerViewContacts.layoutManager = LinearLayoutManager(this@ContactActivity)
recyclerViewContacts.adapter = contactsAdapter
}

cValentyna committed:
level 2 part 2
correction according to the comments in the review.
new corrections

kogorbach commented on app/src/main/java/com/project/myapp/data/contacts/User.kt:
```diff
@@ -0,0 +1,8 @@
+package com.project.myapp.data.contacts
+
+data class User(
+    var id: Int,
```
data classes should only have immutable properties, if you need to change one, you'll have to do something like
```
val newUser = user.copy(it.id = newId)
```

kogorbach commented on app/src/main/java/com/project/myapp/data/contacts/UsersRepository.kt:
```diff
@@ -0,0 +1,72 @@
+
+
+@Singleton
+class UsersRepository @Inject constructor() {
```
Great job on adding a repository, usually repository has an interface, I know it makes little sense when you have only one implementation, but it would be useful to get used to more common architecture and DI structure

cValentyna replied:
added interface IUserRepository and UserRepositoryModule [here](https://github.com/cValentyna/myAndroidProject/pull/18/commits/7d3fc9493597f433757280535317976689a7606a)

kogorbach commented on app/src/main/java/com/project/myapp/screens/contacts/ContactAdapter.kt:
```diff
@@ -0,0 +1,54 @@
+    class ContactViewHolder(
+        private val binding: ContactItemBinding,
+    ) : RecyclerView.ViewHolder(binding.root) {
+        private val currentLibrary = ImageLibrary.GLIDE // can be GLIDE, PICASSO, COIL
```
the implementation of the image loader through interface is really nice, but the general idea of having it, is to declare the library you use for image loading once somewhere in the app, my options are: 
- that exact interface with a default option (could make it a sealed interface since interface + enum is already pretty much teh same approach)
- Application class

cValentyna replied:
image library selection logic was moved into Application class. And I added ImageLoaderModule to create and inject ImageLoader [here](https://github.com/cValentyna/myAndroidProject/pull/18/commits/4df4a6fe5eefa2b54e725f8b80b07c511263e011).

kogorbach commented on app/src/main/java/com/project/myapp/screens/contacts/ContactAdapter.kt:
```diff
@@ -0,0 +1,54 @@
+
+        fun bind(user: User) {
+            binding.contactName.text =
+                buildString {
```
you can do it with
```
"${user.name} ${user.id}"
```

kogorbach commented on app/src/main/java/com/project/myapp/screens/contacts/ContactAdapter.kt:
```diff
@@ -0,0 +1,54 @@
+
+class ContactAdapter(
+    private var contacts: List<User> = emptyList(),
+) : RecyclerView.Adapter<ContactAdapter.ContactViewHolder>() {
```
forgot ListAdapter?

cValentyna replied:
Initially, I chose to implement RecyclerView.Adapter on purpose. But while working on the following task (removing notifyDataSetChanges), I understood  that this approach was not optimal. Your feedback confirmed it, so I switched to ListAdapter and added diffutil [here](https://github.com/cValentyna/myAndroidProject/pull/18/changes/4df4a6fe5eefa2b54e725f8b80b07c511263e011#diff-d566b8c9974b518528a13807632cd02275ea3614a078429de169cffcb9111d87).

cValentyna committed:
level 2 part 1 correction after review
added interface IUserRepository and UserRepositoryModule

cValentyna committed:
level 2 part 1 correction after review

moved image library selection logic into Application class
added ImageLoaderModule to create and inject ImageLoader

Other changes ContactAdapter
replaced RecyclerView.Adapter with ListAdapter
added DiffUtil and removed notifyDataSetChanged()

cValentyna committed:
Merge branch 'dev' into level2_part1_recyclerview
